lib/Serialization/SerializedModuleLoader.cpp

```diff
@@ -1055,6 +1057,17 @@ LoadedFile *SerializedModuleLoaderBase::loadAST(
                        M.getName());
     Ctx.Diags.diagnose(loc, diag::enable_cxx_interop_docs);
   }
+  // Modules built with libc++ cannot be imported into modules that are built
+  // with libstdc++, and vice versa. Make an exception for Cxx.swiftmodule since
+  // it doesn't refer to any C++ stdlib symbols, and for CxxStdlib.swiftmodule
+  // since we skipped loading the overlay for the module.
+  if (M.hasCxxInteroperability() && Ctx.LangOpts.EnableCXXInterop &&
+      M.getCXXStdlibKind() != Ctx.LangOpts.CXXStdlib &&
+      M.getName() != Ctx.Id_Cxx && M.getName() != Ctx.Id_CxxStdlib) {
```

Comment:
What about unconditionally encoding the C++ runtime into the module if C++ Interop is enabled?

Reply:
We are generally doing that, for all modules except Cxx and CxxStdlib. The Cxx module doesn't refer to any C++ stdlib APIs despite being built with C++ interop enabled. The CxxStdlib module has an overlay that is handled separately: ClangImporter skips loading the overlay for CxxStdlib whenever compiling with a non-default C++ stdlib.